### PR TITLE
Added cookie management in retrofit2 and jaxrs clients

### PR DESCRIPTION
**Pull request by mdambros**
Added cookie manager, using configs to set which cookies to accept based on:
- originating host, providing a list of allowed ones
- cookie name, providing a list of regex for the allowed names

---

This change is [Reviewable](https://reviewable.io/reviews/palantir/http-remoting/263)

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/http-remoting](https://github.com/palantir/http-remoting), @mdambros! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

**Comment by mdambros** [comment]
@pnepywoda could you have a look at this? Only a starting point for PDS-46177, but interested to see if this is going in the right direction.

Thanks!

**Comment by mdambros** [comment]
@uschi2000 could you have a look at this PR? The goal is to make http-remoting based services work behind a proxy which requires cookies (as for example a particular configuration of blue coat proxy: https://forums.bluecoat.com/forum/security-policy-enforcement-center/proxysg/30375-bcsi-cookie-on-reverse-proxy)

**Comment by uschi2000** [comment]
Looking

File: jaxrs-clients/build.gradle [hunk 1]
```diff
@@ -22,6 +22,7 @@ dependencies {
     compile "javax.ws.rs:javax.ws.rs-api"
     compile "org.slf4j:slf4j-api"
     compile "com.netflix.feign:feign-slf4j"
+    compile "com.squareup.okhttp3:okhttp-urlconnection:3.0.0-RC1"
```
Comment (uschi2000) [inline jaxrs-clients/build.gradle:25]: Is this versions independently from okhttp3?

File: jaxrs-clients/src/main/java/com/palantir/remoting1/jaxrs/FeignJaxRsClientBuilder.java [hunk 2]
```diff
@@ -69,6 +73,7 @@
 import okhttp3.Credentials;
 import okhttp3.Response;
 import okhttp3.Route;
+import okhttp3.internal.JavaNetCookieJar;
```
Comment (uschi2000) [inline jaxrs-clients/src/main/java/com/palantir/remoting1/jaxrs/FeignJaxRsClientBuilder.java:76]: Is there no public OK http API for cookies?

File: jaxrs-clients/src/main/java/com/palantir/remoting1/jaxrs/FeignJaxRsClientBuilder.java [hunk 3]
```diff
@@ -191,6 +196,36 @@ public long currentTimeMicroseconds() {
...
+            }
+
+            private boolean isAllowedUri(URI uri) {
+                return config.allowedCookieHosts().isPresent()
+                        && config.allowedCookieHosts().get().contains(uri.getHost());
+            }
+
+            private boolean isAllowedCookieName(HttpCookie cookie) {
+                if (config.allowedCookieNameRegex().isPresent()) {
+                    for (String nameRegex : config.allowedCookieNameRegex().get()) {
```
Comment (uschi2000) [inline jaxrs-clients/src/main/java/com/palantir/remoting1/jaxrs/FeignJaxRsClientBuilder.java:215]: Need to precompile these regular expressions into patterns

File: retrofit2-clients/src/main/java/com/palantir/remoting1/retrofit2/Retrofit2ClientBuilder.java [hunk 4]
```diff
@@ -51,6 +57,8 @@ public Retrofit2ClientBuilder(ClientConfig config) {
 
     @Override
     public <T> T build(Class<T> serviceClass, String userAgent, List<String> uris) {
+        CookieManager cm = new CookieManager();
+        CookieHandler.setDefault(cm);
```
Comment (uschi2000) [inline retrofit2-clients/src/main/java/com/palantir/remoting1/retrofit2/Retrofit2ClientBuilder.java:61]: This is a little scary. What does it do and why do we need to set static state?

File: retrofit2-clients/src/main/java/com/palantir/remoting1/retrofit2/Retrofit2ClientBuilder.java [hunk 5]
```diff
@@ -107,6 +115,35 @@ public Request authenticate(Route route, Response response) throws IOException {
         client.readTimeout(config.readTimeout().toMilliseconds(), TimeUnit.MILLISECONDS);
         client.writeTimeout(config.writeTimeout().toMilliseconds(), TimeUnit.MILLISECONDS);
 
+        // cookies
+        CookieManager cm = new CookieManager(null, new CookiePolicy(){
```
Comment (uschi2000) [inline retrofit2-clients/src/main/java/com/palantir/remoting1/retrofit2/Retrofit2ClientBuilder.java:119]: Factor out duplicate code

File: service-config/src/main/java/com/palantir/remoting1/config/service/ServiceConfiguration.java [hunk 6]
```diff
@@ -64,6 +64,17 @@
...
 
+    /**
+     * The list of hosts we are allowed to receive cookies from.
+     */
+    public abstract Optional<List<String>> allowedCookieHosts();
+
+    /**
+     * The list of cookie names we are allowed to receive. It does a cookieName.matches(regex) so regex is allowed
+     */
+    public abstract Optional<List<String>> allowedCookieNameRegex();
```
Comment (uschi2000) [inline service-config/src/main/java/com/palantir/remoting1/config/service/ServiceConfiguration.java:75]: Would prefer allowedCookieNames

**Review by uschi2000 — CHANGES_REQUESTED** [review]

**Comment by uschi2000** [comment]
Also, this needs test coverage.

**Comment by uschi2000** [comment]
Closing since there wasn't much movement

**Pull request closed without merging**